### PR TITLE
Preserve identities of registered task inputs

### LINKED ISSUE #25516: Ad hoc task properties should be implicitly finalized when the configuration cache is used
### Expected Behavior

Ad hoc task properties are implicitly finalized when the configuration cache is used.

### Current Behavior

It looks like at runtime the properties used for the inputs/outputs of the task and the ones referenced in other places are different and not finalized.

### Context (optional)

This was discovered as part of some failing tests in https://github.com/gradle/gradle/pull/25495.

### Steps to Reproduce

See this test: https://github.com/gradle/gradle/pull/25495/files#diff-059f886caa91d008ff4af7b77af2de3aa43d6ee38262e13637846308818f6a79R95

### Gradle version

8.3 nightly

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by bamboo**
Fixes #25516

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test ACC

**Comment by bamboo** [comment]
@bot-gradle merge

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt [hunk 1]
```diff
@@ -294,7 +293,7 @@ suspend fun WriteContext.writeRegisteredPropertiesOf(
         property.run {
             when (this) {
                 is RegisteredProperty.InputFile -> {
-                    val finalValue = DeferredUtil.unpackOrNull(propertyValue)
+                    val finalValue = DeferredUtil.unpackNestableDeferred(propertyValue)
```
Comment (abstratt) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt:296]: Unpack deferreds but do not resolve providers/factories

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt [hunk 2]
```diff
@@ -306,7 +305,7 @@ suspend fun WriteContext.writeRegisteredPropertiesOf(
                 }
 
                 is RegisteredProperty.Input -> {
-                    val finalValue = InputParameterUtils.prepareInputParameterValue(propertyValue)
+                    val finalValue = DeferredUtil.unpackNestableDeferred(propertyValue)
```
Comment (abstratt) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt:308]: Unpack deferreds but do not resolve providers/factories, nor coherce `GString` into `String`

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDependencyInferenceIntegrationTest.groovy [hunk 3]
```diff
@@ -929,7 +929,6 @@ The following types/formats are supported:
         file("out.txt").text == "22"
     }
 
-    @ToBeFixedForConfigurationCache(because = "queries mapped value of task output before it has completed")
     def "ad hoc input property with value of mapped task output implies dependency on the task"() {
```
Comment (abstratt) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDependencyInferenceIntegrationTest.groovy:932]: 🎉

**Review by abstratt — APPROVED** [review]
LGTM!

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt (hunk 2 above)
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt:308]: Yes, coercion from `GString` to `String` should happen at a different level uniformly for all properties.

**Comment by bamboo** [comment]
@bot-gradle merge

**Comment by bamboo** [comment]
@bot-gradle merge

**Pull request merged**